flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/ZooKeeperSubmittedJobGraphs.java

```diff
@@ -0,0 +1,356 @@
+
+		// Ensure that the job graphs path exists
+		client.newNamespaceAwareEnsurePath(currentJobsPath)
+				.ensure(client.getZookeeperClient());
+
+		// All operations will have the path as root
+		client = client.usingNamespace(client.getNamespace() + currentJobsPath);
+
+		this.jobGraphsInZooKeeper = new ZooKeeperStateHandleStore<>(client, stateHandleProvider);
+
+		this.pathCache = new PathChildrenCache(client, "/", false);
+		pathCache.getListenable().addListener(new SubmittedJobGraphsPathCacheListener());
+	}
+
+	@Override
+	public void start(SubmittedJobGraphListener jobGraphListener) throws Exception {
+		synchronized (cacheLock) {
+			if (!isRunning) {
+				this.jobGraphListener = jobGraphListener;
+
+				pathCache.start();
+
+				isRunning = true;
+			}
+		}
+	}
+
+	@Override
+	public void stop() throws Exception {
+		synchronized (cacheLock) {
+			if (isRunning) {
+				jobGraphListener = null;
+
+				pathCache.close();
+
+				client.close();
+
+				isRunning = false;
+			}
+		}
+	}
+
+	@Override
+	public List<SubmittedJobGraph> recoverJobGraphs() throws Exception {
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			List<ZooKeeperStateHandle<SubmittedJobGraph>> submitted;
+
+			while (true) {
+				try {
+					submitted = jobGraphsInZooKeeper.getAll();
+					break;
+				}
+				catch (ConcurrentModificationException e) {
+					LOG.warn("Concurrent modification while reading from ZooKeeper. Retrying.");
+				}
+			}
+
+			if (submitted.size() != 0) {
+				List<SubmittedJobGraph> jobGraphs = new ArrayList<>(submitted.size());
+
+				for (ZooKeeperStateHandle<SubmittedJobGraph> jobStateHandle : submitted) {
+					SubmittedJobGraph jobGraph = jobStateHandle
+							.getStateHandle()
+							.getState(ClassLoader.getSystemClassLoader());
+
+					addedJobGraphs.add(jobGraph.getJobId());
+
+					jobGraphs.add(jobGraph);
+				}
+
+				LOG.info("Recovered {} job graphs: {}.", jobGraphs.size(), jobGraphs);
+				return jobGraphs;
+			}
+			else {
+				LOG.info("No job graph to recover.");
+				return Collections.emptyList();
+			}
+		}
+	}
+
+	@Override
+	public SubmittedJobGraph recoverJobGraph(JobID jobId) throws Exception {
+		checkNotNull(jobId, "Job ID");
+		String path = getPathForJob(jobId);
+
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			StateHandle<SubmittedJobGraph> jobStateHandle = jobGraphsInZooKeeper.get(path);
+
+			SubmittedJobGraph jobGraph = jobStateHandle
+					.getState(ClassLoader.getSystemClassLoader());
+
+			addedJobGraphs.add(jobGraph.getJobId());
+
+			LOG.info("Recovered {}.", jobGraph);
+			return jobGraph;
+		}
+	}
+
+	@Override
+	public void putJobGraph(SubmittedJobGraph jobGraph) throws Exception {
+		checkNotNull(jobGraph, "Job graph");
+		String path = getPathForJob(jobGraph.getJobId());
+
+		synchronized (cacheLock) {
+			verifyIsRunning();
+
+			int currentVersion = jobGraphsInZooKeeper.exists(path);
+
+			if (currentVersion == -1) {
+				jobGraphsInZooKeeper.add(path, jobGraph);
```

Comment:
After thinking about it, I think it's better to fail the submission. How often would you retry in case of permanent problem? If the submission fails, the user can react to it and we get a clear error. I might also be wrong. I feel like this is something we should leave for later.